Descriptions/Comments/Locations rework

sungam3r opened the pull request:
See #132 as a starting point. Fixes #141. Also now parser respects _ignore-locations_ option for comments. Also fixes exceptions caused by placing comments in "unexpected" positions in query/SDL.

sungam3r commented on .github/workflows/test.yml:
```diff
@@ -41,7 +41,7 @@ jobs:
         uses: actions/setup-dotnet@v1
         with:
-          dotnet-version: '5.0.201'
+          dotnet-version: '5.0.x'
```
@Shane32 Do you remember why we install .NET Core 3.1 SDK ?

Shane32 replied:
Tests run against both 3.1 and 5.0. I wouldn’t change that as 5.0 is not a LTS version.

Shane32 commented on src/GraphQLParser/ParserContext.Parse.cs:
```diff
@@ -250,7 +249,7 @@ private GraphQLDirectiveDefinition ParseDirectiveDefinition()
             var locations = ParseDirectiveLocations();
 
-            return _ignoreOptions == IgnoreOptions.IgnoreCommentsAndLocations
+            return _ignoreOptions.HasFlag(IgnoreOptions.Locations)
```
All of these conditions are wrong.  If only 'Locations' is specified to be skipped, 'Comments' are also skipped because the non-"full" version is used which does not have a field to store the comments node.

```suggestion
            return _ignoreOptions == IgnoreOptions.All
```

Shane32 commented on src/GraphQLParser/ParserContext.Parse.cs:
```diff
@@ -250,7 +249,7 @@ private GraphQLDirectiveDefinition ParseDirectiveDefinition()
             var locations = ParseDirectiveLocations();
 
-            return _ignoreOptions == IgnoreOptions.IgnoreCommentsAndLocations
+            return _ignoreOptions.HasFlag(IgnoreOptions.Locations)
```
Many of these conditions are wrong.  For instance, here if only 'Locations' is specified to be skipped, 'Comments' are also skipped because the non-"full" version is used which does not have a field to store the comments node.

```suggestion
            return _ignoreOptions == IgnoreOptions.All
```

Shane32 replied:
Tests should be added to ensure that each combination of ignored values is properly ignored or retained.

Shane32 replied:
I think actually the problem isn't that the tests don't exist, but that no `[InlineData(IgnoreOptions.Locations)]` were added to tests for comments

sungam3r replied:
Well, my mistake is obvious 🤦 .

Shane32 commented on src/GraphQLParser/AST/GraphQLComment.cs:
```diff
@@ -17,17 +16,22 @@ public class GraphQLComment : ASTNode
         //}
     }
+
+    internal class GraphQLCommentFull : GraphQLComment
+    {
+        private GraphQLLocation _location;
+
+        public override GraphQLLocation Location
+        {
+            get => _location;
+            set => _location = value;
+        }
+    }
```
Nowhere else do we have a special class for comments without a location.  Optimization is only for ignoring both comments and locations.

sungam3r replied:
I renamed `GraphQLCommentFull` into `GraphQLCommentWithLocation`. I will soon make a commit, where I added a lot of new internal classes to align the behavior. All optimizations will be in place. I will need some time to finish and write tests. I want to show in advance what it turns out.

Shane32 commented on src/GraphQLParser/IgnoreOptions.cs:
```diff
@@ -1,23 +1,31 @@
 
         /// <summary>
-        /// Specifies whether to ignore comments and token locations when parsing GraphQL document.
+        /// Specifies whether to ignore  token locations when parsing GraphQL document.
```
```suggestion
        /// Specifies whether to ignore token locations when parsing GraphQL document.
```

sungam3r replied:
👍

Shane32 commented on src/GraphQLParser/AST/GraphQLTypeDefinition.cs:
```diff
@@ -1,16 +1,11 @@
         public GraphQLName? Name { get; set; }
-    }
 
-    public abstract class GraphQLTypeDefinitionWithDescription : GraphQLTypeDefinition, IHasDescription
```
Ok to remove.

Shane32 commented on src/GraphQLParser.Tests/ParserTests.cs:
```diff
@@ -1082,9 +1144,14 @@ enum Colors {
             var defs = document.Definitions;
-            defs.Count.ShouldBe(7);
+            defs.Count.ShouldBe(8);
             var parseComments = options == IgnoreOptions.None;
```
```suggestion
            var parseComments = options == IgnoreOptions.None || options == IgnoreOptions.Locations;
```

Shane32 replied:
You may want to check to see a similar change is needed anywhere else in the tests.

Shane32 replied:
Or this

```suggestion
            var parseComments = !options.HasFlag(IgnoreOptions.Comments);
```

Shane32 commented on src/GraphQLParser.Tests/ParserTests.cs:
```diff
@@ -910,9 +960,14 @@ enum Colors {
             var defs = document.Definitions;
-            defs.Count.ShouldBe(7);
+            defs.Count.ShouldBe(8);
             var parseComments = options == IgnoreOptions.None;
```
```suggestion
            var parseComments = !options.HasFlag(IgnoreOptions.Comments);
```

Shane32 commented on src/GraphQLParser.Tests/GraphQLAstVisitorTests.cs:
```diff
@@ -54,8 +54,8 @@ public GraphQLAstVisitorTests()
         [InlineData(IgnoreOptions.None)]
-        [InlineData(IgnoreOptions.IgnoreComments)]
-        [InlineData(IgnoreOptions.IgnoreCommentsAndLocations)]
+        [InlineData(IgnoreOptions.Comments)]
+        [InlineData(IgnoreOptions.All)]
```
```suggestion
        [InlineData(IgnoreOptions.Comments)]
        [InlineData(IgnoreOptions.Locations)]
        [InlineData(IgnoreOptions.All)]
```
For the sake of consistency, shouldn't Locations be here also or am I missing something?  Also line 69

sungam3r replied:
Yes.

Shane32 commented on src/GraphQLParser.Tests/ParserTests.cs:
```diff
@@ -47,8 +47,9 @@ public void Extra_Comments_Should_Read_Correctly(IgnoreOptions options)
         //[InlineData(IgnoreOptions.None)]
-        [InlineData(IgnoreOptions.IgnoreComments)]
-        [InlineData(IgnoreOptions.IgnoreCommentsAndLocations)]
+        [InlineData(IgnoreOptions.Comments)]
```
Changes should be made on various tests (e.g. the `Comments_on_FragmentSpread_Should_Read_Correclty` test) to include `[InlineData(IgnoreOptions.Locations)]` to test reading comments when only locations are ignored

Shane32 replied:
(Not this test, but I couldn't comment on line 69)

sungam3r replied:
Yes

Shane32 commented:
Note that GraphQLLocation is a struct, and takes almost no processing power to compute (since the location is already known).  So if there are no optimized classes for classes, then running the parser with only `IgnoreOptions.Locations` is pointless, as it will use the same memory, and probably an identical amount of processing time.  The reason the option existed previously is because when ignoring both comments and locations (neither of which are necessary for execution of a query by GraphQL.NET), optimized classes can be used that save memory requirements of both fields, and the heap allocations and processing power to process comment fields.

I also think (if I remember right) that if an error occurs during parsing, an exception is generated within the parser that utilizes the location of the error, even if location is specified to be ignored when parsing.

So while flags are a better design, it may be a bit misleading to users in that ignoring locations (but not comments) will actually do anything beneficial.  Just because of that reason, I'd probably keep the options values like they were, especially as we have no intention of adding additional flags or adding additional classes to options for ignoring only locations.  Assume that we proceed with switching it to flags, I might add a comment to this effect - either in the xml comment for the enum option, or in the readme file.

sungam3r commented:
>  So if there are no optimized classes for skipping only locations, then running the parser with only IgnoreOptions.Locations is pointless, as it will use the same memory, and probably an identical amount of processing time
> So while flags are a better design, it may be a bit misleading to users in that ignoring locations (but not comments) will actually do anything beneficial.

I did so that every option gave effect. I have to write tests and put some code for reading comments in 3-4 places. When I worked on this PR, I realized that the processing of comments is a very unpleasant process. They may be everywhere! We are far from properly handling comment. I found that if you put a comment after the name of the arg but before its value, the parser generates an exception. Boom! See `CommentsOnValues.graphql` file. I have to add `ParseComment()` call into more places.

sungam3r commented on src/GraphQLParser/NodeHelper.cs:
```diff
@@ -0,0 +1,416 @@
+    internal static class NodeHelper
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static GraphQLComment CreateGraphQLComment(IgnoreOptions options)
```
First three methods differ from others. `GraphQLComment`, `GraphQLDocument` and `GraphQLDescription` have no comments. All other methods are completely the same. Perhaps when I will write tests, I will see that comments are not applicable to some other AST nodes.

Shane32 commented:
Why not just discard all comments (according to spec) for v8?  Eliminate the options for parsing it, eliminate the complexity of the classes to hold comments, etc.

> Comments are Ignored like white space and may appear after any token, or before a LineTerminator, and have no significance to the semantic meaning of a GraphQL Document.

Shane32 commented:
> I renamed GraphQLCommentFull into GraphQLCommentWithLocation. I will soon make a commit, where I added a lot of new internal classes to align the behavior. All optimizations will be in place. I will need some time to finish and write tests. I want to show in advance what it turns out.

I really think you are wasting a lot of time here, and making the project unnecessarily complex with "a lot of new internal classes" that are not necessary when the only project that uses this parser -- GraphQL.NET -- only does two things: (a) skip locations & comments for execution, and (b) load comments for backwards compatibility when creating a schema-first schema.  And as for the second point, descriptions have long been in the spec now to be done by strings, not comments.  So even that is not necessary.  And even if 'comments without locations' could be used by the schema-first initialization, memory requirements does not matter because schema initialization is only done once.  (Or should be.)

Shane32 commented:
(Edit: oops GraphQL always uses locations)

We should really eliminate parsing without locations, since GraphQL requires the use of locations in error messages.

sungam3r commented:
> Why not just discard all comments (according to spec) for v8? Eliminate the options for parsing it, eliminate the complexity of the classes to hold comments, etc.

Of course, we **can** do it. But from the point of view of the syntactic parsing it will be a step back. People may need comments on completely different reasons. The most obvious of them is the documentation provision through the introspection. We will force everyone to use descriptions. At best, consumers will have to change their SDLs. In the worst case, they will not be able to do this at all, because they do not own these schemas. Then they have to change these schemas (SDL) dynamically, which will require  ... the syntactic analysis that we have deleted 😄 .

sungam3r commented:
> We should really eliminate optimization of parsing without locations, since GraphQL requires the use of locations in error messages.

I'm not sure that this requirement somehow limits us. The specification describes what should occur between the client and the server. Parser is a general mechanism that can work without the presence of any client. For example, it may be the transformation from one SDL to another SDL or to some different file format. Getting SDL statistics, whatever.

Shane32 commented:
None of those scenarios require memory optimization.  Only execution does, which requires location information.

sungam3r commented:
> when the only project that uses this parser -- GraphQL.NET 

Why so 😄 ? It is better to say "the only **public** project that uses this parser" although I doubt even that it is indeed the only public. I use the parser in my private projects for example. In any case, we need to eliminate exceptions caused by the placing comments in "unexpected" positions.

Shane32 commented:
You can write all that extra code if you want; I wouldn't 🤷‍♂️.

Shane32 commented:
> In any case, we need to eliminate exceptions caused by the placing comments in "unexpected" positions.

For sure.

Shane32 commented:
> I use the parser in my private projects for example.

Where (a) you skip parsing location information, and (b) in situations that require hyper-optimized memory allocations?

Shane32 commented:
Hey, go ahead 👍 ... maybe it's useful to someone; maybe you.

sungam3r commented:
OK. By the way, if we talk about the complexity of the code, then after the changes made code became easier. I deleted all the ternary operators around AST creation and initializing. All properties are set once now. The number of codes in the "main" parser file decreased. I look positively on changes.

Shane32 commented:
Is it ready for another review?

sungam3r commented:
Not yet. I'll let you know.

sungam3r commented:
@Shane32 I'm ready for review. It turned out to be a big one PR.

What you should pay attention to:
1. Single `ParseComment()` call inside `Advance` so now we handle all comments anywhere.
2. All nodes parse and set their comments now.
3. New tests verify additional comments in new places. I can add additional tests for comments in awkward positions like before `:` or `=` tokens. Now all of them should reside into unattached comments. No more exceptions!
4. `ParserContext` ctor calls `CreateGraphQLDocument` and `Advance`.
5. I renamed some public classes to align their names to ones from the spec: GraphQLFieldSelection -> GraphQLField, GraphQLOperationTypeDefinition  -> GraphQLRootOperationTypeDefinition

I also walked through all the `ParseXXX` methods and added comment to each with reference to the corresponding part of the specification (grammar).

sungam3r commented:
Now it will be easier to work on the new features from the October spec.
